# NLP Assignment

# Group No - 20

## Group Member Names:
1. Varinder Singh - 2021fc04070@wilp.bits-pilani.ac.in
2. Bandaru Raja Sekhar - 2021fc04074@wilp.bits-pilani.ac.in
3. Mikhil P.A. - 2021fc04326@wilp.bits-pilani.ac.in
4. Yogesh Tiwari - 2021fc04825@wilp.bits-pilani.ac.in

# Problem Statement - 42
 
## Set 8

Amazon Alexa Reviews from URL: https://www.kaggle.com/datasets/sid321axn/amazon-alexa-reviews

This dataset consists of nearly 3000 Amazon customer reviews (input text), star ratings, date of review, variants, and feedback of various amazon Alexa products like Alexa Echo, Echo dots, Alexa Firesticks, etc. for learning how to train Machine for sentiment analysis.

 

# Point 1


### Read the Amazon.csv file and set it as a Data frame called Amazon. Describe the Data set head, No. of rows, columns, and attributes. (3 Marks)

In [1]:
!pip install nltk
!pip install spacy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 33.9 MB/s eta 0:00:00


In [2]:
import nltk
import numpy as np
import pandas as pd
dataset_tsv = pd.read_csv("amazon_alexa.tsv", sep='\t')

In [3]:
dataset = dataset_tsv.copy()

In [4]:
dataset_tsv.head()

rating 
 date 
 variation 
 verified_reviews 
 feedback 
 
 
 
 
 0 
 5 
 31-Jul-18 
 Charcoal Fabric 
 Love my Echo! 
 1 
 
 
 1 
 5 
 31-Jul-18 
 Charcoal Fabric 
 Loved it! 
 1 
 
 
 2 
 4 
 31-Jul-18 
 Walnut Finish 
 Sometimes while playing a game, you can answer... 
 1 
 
 
 3 
 5 
 31-Jul-18 
 Charcoal Fabric 
 I have had a lot of fun with this thing. My 4 ... 
 1 
 
 
 4 
 5 
 31-Jul-18 
 Charcoal Fabric 
 Music 
 1

### No. of rows, columns, and Column list

In [5]:
print('No. of Rows: ', len(dataset))
print('No. of Columns: ', len(dataset.columns))
print('Column list: ', list(dataset.columns))

No. of Rows:  3150
No. of Columns:  5
Column list:  ['rating', 'date', 'variation', 'verified_reviews', 'feedback']


### Dataset Info

In [6]:
print("Dataset Shape:", dataset.shape, "\n")
dataset.info()

Dataset Shape: (3150, 5) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3150 entries, 0 to 3149
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   rating            3150 non-null   int64 
 1   date              3150 non-null   object
 2   variation         3150 non-null   object
 3   verified_reviews  3150 non-null   object
 4   feedback          3150 non-null   int64 
dtypes: int64(2), object(3)
memory usage: 123.2+ KB


# Point 2

### Remove punctuations and stop words from the text in the ‘text’ column (2Marks)

In [7]:
from nltk import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package wordnet to /home/datalore/nltk_data...
[nltk_data] Downloading package stopwords to
[nltk_data]     /home/datalore/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /home/datalore/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [8]:
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    tokens = word_tokenize(text.lower())
    tokens = [lemmatizer.lemmatize(token) for token in tokens if token.isalpha() and token not in stop_words]
    return ' '.join(tokens)

dataset['verified_reviews'] = dataset['verified_reviews'].apply(preprocess_text)

In [9]:
# Displaying few rows from dataset 

dataset.head()

rating 
 date 
 variation 
 verified_reviews 
 feedback 
 
 
 
 
 0 
 5 
 31-Jul-18 
 Charcoal Fabric 
 love echo 
 1 
 
 
 1 
 5 
 31-Jul-18 
 Charcoal Fabric 
 loved 
 1 
 
 
 2 
 4 
 31-Jul-18 
 Walnut Finish 
 sometimes playing game answer question correct... 
 1 
 
 
 3 
 5 
 31-Jul-18 
 Charcoal Fabric 
 lot fun thing yr old learns dinosaur control l... 
 1 
 
 
 4 
 5 
 31-Jul-18 
 Charcoal Fabric 
 music 
 1

# Point 3

### Create two objects X and y. X will be the 'text' column of Dataset and y will be yelp's 'stars' column. create a CountVectorizer object and split the data into training and testing sets. Train a MultinomialNB model and Display the confusion Matrix (5 Marks)

In [10]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import confusion_matrix

In [11]:
X = dataset['verified_reviews']
y = dataset['rating']

#### CountVectorizer 
CountVectorizer is a tool used in natural language processing to convert a collection of text documents into a matrix of token counts. It is a type of feature extraction technique used in machine learning, where each row of the matrix represents a document and each column represents a specific word in the vocabulary. The value in each cell of the matrix is the count of how many times that word appears in that particular document. This matrix can then be used as input for various machine learning algorithms.

In [12]:
cv = CountVectorizer()

In [13]:
X = cv.fit_transform(X)

#### Splitting the dataset into Training and Testing dataset

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

#### MultinomialNB (Multinomial Naive Bayes) 
MultinomialNB (Multinomial Naive Bayes) is a type of Naive Bayes algorithm that is commonly used for text classification tasks, such as spam filtering or sentiment analysis. One of the main reasons to use MultinomialNB is its simplicity and efficiency. It is a fast and easy-to-implement algorithm that can handle large datasets with many features (such as the large number of words in a text corpus) and is less prone to overfitting compared to other more complex models.

In [15]:
nb = MultinomialNB()
nb.fit(X_train, y_train)

MultinomialNB()

In [16]:
y_pred = nb.predict(X_test)

In [17]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[ 12   2   1   5  13]
 [  1   0   0   8  12]
 [  1   0   6  12  20]
 [  1   0   0  15  74]
 [  1   1   0  13 432]]


This confusion matrix represents the performance of a classifier model that was used to predict. The rows of the matrix represent the actual ratings given by the reviewers, while the columns represent the predicted ratings made by the classifier. The entries in the matrix count the number of reviews that fall into each category.

For example, the entry in the first row and first column (12) represents the number of reviews that were actually rated with 1 star and were correctly predicted to have 1 star by the model. The entry in the second row and fourth column (8) represents the number of reviews that were actually rated with 4 stars but were mistakenly predicted to have 2 stars by the model. Overall, the confusion matrix shows that the model performs well for reviews with a 5-star rating (with 432 correct predictions out of 447), but performs poorly for reviews with a 1-star rating (with only 12 correct predictions out of 33).

In [18]:
# Calculating the accuracy score of the model
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)

Accuracy: 0.7380952380952381


# Point 4

### Display the HMM POS tagging on the first 4 rows of ‘text’ (3 Marks)

HMM POS tagging refers to the use of Hidden Markov Models (HMMs) for part-of-speech (POS) tagging in natural language processing. POS tagging involves assigning a grammatical tag to each word in a sentence, indicating its part of speech, such as noun, verb, adjective, etc. HMMs are statistical models that can capture the probability distribution of a sequence of observed events. In the case of POS tagging, the observed events are the words in a sentence, and the hidden events are the POS tags.

In [19]:
# Downloading the packages from nltk
nltk.download('averaged_perceptron_tagger')
nltk.download('universal_tagset')
nltk.download('tagsets')

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/datalore/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package universal_tagset to
[nltk_data]     /home/datalore/nltk_data...
[nltk_data]   Unzipping taggers/universal_tagset.zip.
[nltk_data] Downloading package tagsets to /home/datalore/nltk_data...
[nltk_data]   Unzipping help/tagsets.zip.


True

In [20]:
#POS tagging of data from tsv files

for i in range(4):
    text = dataset_tsv['verified_reviews'][i]
    print(text.lower())
    tokens = nltk.word_tokenize(text.lower())
    pos_tags = nltk.pos_tag(tokens, tagset='universal')
    print(pos_tags)
    print()

love my echo!
[('love', 'VERB'), ('my', 'PRON'), ('echo', 'NOUN'), ('!', '.')]

loved it!
[('loved', 'VERB'), ('it', 'PRON'), ('!', '.')]

sometimes while playing a game, you can answer a question correctly but alexa says you got it wrong and answers the same as you.  i like being able to turn lights on and off while away from home.
[('sometimes', 'ADV'), ('while', 'ADP'), ('playing', 'VERB'), ('a', 'DET'), ('game', 'NOUN'), (',', '.'), ('you', 'PRON'), ('can', 'VERB'), ('answer', 'VERB'), ('a', 'DET'), ('question', 'NOUN'), ('correctly', 'ADV'), ('but', 'CONJ'), ('alexa', 'VERB'), ('says', 'VERB'), ('you', 'PRON'), ('got', 'VERB'), ('it', 'PRON'), ('wrong', 'ADJ'), ('and', 'CONJ'), ('answers', 'VERB'), ('the', 'DET'), ('same', 'ADJ'), ('as', 'ADP'), ('you', 'PRON'), ('.', '.'), ('i', 'VERB'), ('like', 'ADP'), ('being', 'VERB'), ('able', 'ADJ'), ('to', 'PRT'), ('turn', 'VERB'), ('lights', 'NOUN'), ('on', 'ADP'), ('and', 'CONJ'), ('off', 'ADP'), ('while', 'ADP'), ('away', 'ADV'), ('from

In [21]:
#POS tagging of processed data from point 3

for i in range(4):
    text = dataset['verified_reviews'][i]
    print(text)
    tokens = nltk.word_tokenize(text)
    pos_tags = nltk.pos_tag(tokens, tagset='universal')
    print(pos_tags)
    print()

love echo
[('love', 'NOUN'), ('echo', 'NOUN')]

loved
[('loved', 'VERB')]

sometimes playing game answer question correctly alexa say got wrong answer like able turn light away home
[('sometimes', 'ADV'), ('playing', 'VERB'), ('game', 'NOUN'), ('answer', 'ADJ'), ('question', 'NOUN'), ('correctly', 'ADV'), ('alexa', 'ADJ'), ('say', 'VERB'), ('got', 'VERB'), ('wrong', 'ADJ'), ('answer', 'NOUN'), ('like', 'ADP'), ('able', 'ADJ'), ('turn', 'NOUN'), ('light', 'ADJ'), ('away', 'ADV'), ('home', 'NOUN')]

lot fun thing yr old learns dinosaur control light play game like category nice sound playing music well
[('lot', 'NOUN'), ('fun', 'ADJ'), ('thing', 'NOUN'), ('yr', 'NOUN'), ('old', 'ADJ'), ('learns', 'VERB'), ('dinosaur', 'ADJ'), ('control', 'NOUN'), ('light', 'ADJ'), ('play', 'NOUN'), ('game', 'NOUN'), ('like', 'ADP'), ('category', 'NOUN'), ('nice', 'ADJ'), ('sound', 'NOUN'), ('playing', 'NOUN'), ('music', 'NOUN'), ('well', 'NOUN')]



# Point 5

### Parse the first 4 rows of ‘text’ using Viterbi Parser ( [Use toy_pcfg1 and toy_pcfg2 to get the probabilistic context-free grammars; use the PCFG suitable for each sentence] (2 marks)

In [22]:
from functools import reduce
from nltk.grammar import PCFG
from nltk.parse import ViterbiParser

PCFG stands for Probabilistic Context-Free Grammar, which is a formalism used in natural language processing (NLP) to describe the syntax or structure of a language. It is a type of context-free grammar where each production rule has a probability associated with it, which represents the likelihood of that rule being used in a sentence.

PCFGs are useful in NLP for several reasons:

- **Parsing**: One important application of PCFGs is in parsing, where the goal is to automatically analyze the structure of a sentence and identify the relationships between its words. The probabilities associated with each rule in a PCFG can be used to guide the parsing process, helping to select the most likely parse tree for a given sentence.

- **Language modeling**: PCFGs can also be used to model the probability distribution over sentences in a language. This can be used to generate new sentences or to evaluate the likelihood of a given sentence being grammatically correct.

- **Machine translation**: PCFGs can be used in machine translation systems to help align the structure of the source and target languages, which can improve the accuracy of the translation.


In [23]:
toy_pcfg1 = PCFG.fromstring(
"""
    S -> VP NP [0.5] | VP [0.5]
    VP -> V [0.5] | V Det [0.5]
    V -> 'love' [0.5] | 'loved' [0.5]
    NP -> Det N [1]
    Det -> 'my' [0.5] | 'it' [0.5]
    N -> 'echo' [1]
""")

In [24]:
toy_pcfg2 = PCFG.fromstring(
"""
    S -> ADVP NP1 CONJ NP2 CONJ NP3 PNP CONJ NP4 [0.5] | S1 S2 S3 [0.5]
    S1 -> PN V V DET N ADP N ADP DET N [1.0]
    S2 -> NP5  NP6 CONJ NP7 [1.0]
    S3 -> NP8 ADV NP9 [1.0]
    NP5 -> PN NUM ADV ADJ V ADP N [1.0]
    NP6 -> PN V DET N [1.0]
    NP7 -> V N ADP N [1.0]
    NP8 -> V ADJ N [1.0]
    NP9 -> V N ADV ADV [1.0]
    ADVP -> ADV ADP V DET N  [1.0]
    NP1 -> PN V V DET N ADV [1.0] 
    NUM -> '4' [1.0]
    NP2 -> N VNP VNP ADJ [1.0] 
    VNP -> V PN [1.0]
    NP3 -> V DET ADJ ADP PN [1.0] 
    VP -> V V V [1.0]
    NP4 -> ADP ADP ADV ADP N [1.0]
    PNP -> PN VP PRT V N ADP [1.0]
    AD -> ADP ADP [1]
    PRT -> 'to' [1.0]
    CONJ -> 'and' [0.7] | 'but' [0.3]
    N -> 'lot' [0.05] | 'games' [0.1] | 'fun' [0.05] | 'thing' [0.1] | 'dinosaurs' [0.05] | 'music' [0.1] | 'categories' [0.05] | 'sound' [0.05] | 'game' [0.05] | 'question' [0.1] | 'alexa' [0.1] | 'lights' [0.1] | 'home' [0.1]
    PN -> 'my' [0.1] | 'you' [0.7] | 'it' [0.1] | 'i' [0.1]
    ADJ -> 'old' [0.25] | 'nice' [0.25] | 'wrong' [0.25] | 'same' [0.25]
    ADV -> 'well' [0.1] | 'when' [0.2] | 'yr' [0.1] | 'sometimes' [0.2] | 'correctly' [0.2] | 'away' [0.2]
    ADP -> 'of' [0.1] | 'with' [0.1] | 'about' [0.1] | 'while' [0.3] | 'as' [0.1] | 'on' [0.1] | 'from' [0.1] | 'off' [0.1]
    V -> 'control' [0.025] | 'have' [0.05] | 'had' [0.05] | 'learns' [0.05] | 'play' [0.125] | 'has' [0.05] | 'playing'  [0.1]  | 'can'  [0.05] | 'answer'  [0.1]  | 'says' [0.1] | 'got' [0.05] | 'answers' [0.05] | 'like' [0.05] | 'being' [0.05] | 'able' [0.05] | 'turn' [0.05]  
    DET -> 'a' [0.6] | 'the' [0.3] | 'this' [0.1] | 
""")

### Getting first four sentences from the dataset

In [25]:
import string

# Code to remove punctuations from sentences
def remove_punc(sentence):
    translator = str.maketrans('', '', string.punctuation)
    sentence_without_punct = sentence.translate(translator)
    return sentence_without_punct.lower()

sentences = [
        (remove_punc(dataset_tsv["verified_reviews"][0]), toy_pcfg1),
        (remove_punc(dataset_tsv["verified_reviews"][1]), toy_pcfg1),
        (remove_punc(dataset_tsv["verified_reviews"][2]), toy_pcfg2),
        (remove_punc(dataset_tsv["verified_reviews"][3]), toy_pcfg2),
    ]
sentences

[('love my echo', <Grammar with 10 productions>),
 ('loved it', <Grammar with 10 productions>),
 ('sometimes while playing a game you can answer a question correctly but alexa says you got it wrong and answers the same as you  i like being able to turn lights on and off while away from home',
  <Grammar with 78 productions>),
 ('i have had a lot of fun with this thing my 4 yr old learns about dinosaurs i control the lights and play games like categories has nice sound when playing music as well',
  <Grammar with 78 productions>)]

In [26]:
# Function to print some summary statistics

def print_stats(all_parses, num_parses):
    print()
    print("Time (secs)   # Parses   Average P(parse)")
    print("-----------------------------------------")
    print("%11.4f%11d%19.14f" % (time, num_parses, average))
    parses = all_parses.keys()
    if parses:
        p = reduce(lambda a, b: a + b.prob(), parses, 0) / len(parses)
    else:
        p = 0
    print("------------------------------------------")
    print("%11s%11d%19.14f" % ("n/a", len(parses), p))

    if num_parses > 0 :
        print("\n\n------------------------------------")
        print("Printing parses")
        print("------------------------------------")
        for parse in parses:
            print(parse)

## Taking Row 1 from dataset

In [27]:
index = 0
sent, grammar = sentences[index]
sent

'love my echo'

**Viterbi parser** is a type of algorithm used in natural language processing (NLP) to analyze and parse sentences by assigning a probability to each possible parse tree of the input sentence, and then selecting the parse tree with the highest probability as the correct one.

The Viterbi algorithm is a dynamic programming algorithm that recursively calculates the probability of each possible parse tree of a sentence, given a probabilistic context-free grammar (PCFG) and a set of observed words. It is based on the assumption that the probability of a parse tree can be calculated as the product of the probabilities of its constituent parts (i.e., the probability of a subtree is the product of the probabilities of its children).

Viterbi parser is used in various NLP tasks such as part-of-speech tagging, named entity recognition, and syntactic parsing. It is particularly useful in cases where the input sentence is ambiguous and can have multiple possible parse trees. By assigning probabilities to each possible parse tree, Viterbi parser can select the most likely interpretation of the sentence, which is essential for many NLP applications such as machine translation, speech recognition, and information retrieval.

In [28]:
# Tokenize the sentence.
tokens = sent.split()

parser = ViterbiParser(grammar)
all_parses = {}

print("---------------------------------------------------------")
print(f"\nsentence: {sent}\nparser: {parser}\ngrammar: {grammar}")
print("---------------------------------------------------------\n\n\n")

---------------------------------------------------------

sentence: love my echo
parser: <ViterbiParser for <Grammar with 10 productions>>
grammar: Grammar with 10 productions (start state = S)
    S -> VP NP [0.5]
    S -> VP [0.5]
    VP -> V [0.5]
    VP -> V Det [0.5]
    V -> 'love' [0.5]
    V -> 'loved' [0.5]
    NP -> Det N [1.0]
    Det -> 'my' [0.5]
    Det -> 'it' [0.5]
    N -> 'echo' [1.0]
---------------------------------------------------------





In [29]:
import time

parser.trace(3)
t = time.time()
parses = parser.parse_all(tokens)
time = time.time() - t
average = (
    reduce(lambda a, b: a + b.prob(), parses, 0) / len(parses) if parses else 0
)

Inserting tokens into the most likely constituents table...
   Insert: |=..| love
   Insert: |.=.| my
   Insert: |..=| echo
Finding the most likely constituents spanning 1 text elements...
   Insert: |=..| V -> 'love' [0.5]                  0.5000000000 
   Insert: |=..| VP -> V [0.5]                      0.2500000000 
   Insert: |=..| S -> VP [0.5]                      0.1250000000 
   Insert: |.=.| Det -> 'my' [0.5]                  0.5000000000 
   Insert: |..=| N -> 'echo' [1.0]                  1.0000000000 
Finding the most likely constituents spanning 2 text elements...
   Insert: |==.| VP -> V Det [0.5]                  0.1250000000 
   Insert: |==.| S -> VP [0.5]                      0.0625000000 
   Insert: |.==| NP -> Det N [1.0]                  0.5000000000 
Finding the most likely constituents spanning 3 text elements...
   Insert: |===| S -> VP NP [0.5]                   0.0625000000 


In [30]:

num_parses = len(parses)
print("Number of parses: ", num_parses)

for p in parses:
    all_parses[p.freeze()] = 1

print("---------------------------------------------------------")
print_stats(all_parses, num_parses)
index = index + 1

Number of parses:  1
---------------------------------------------------------

Time (secs)   # Parses   Average P(parse)
-----------------------------------------
     0.0198          1   0.06250000000000
------------------------------------------
        n/a          1   0.06250000000000


------------------------------------
Printing parses
------------------------------------
(S (VP (V love)) (NP (Det my) (N echo))) [0.0625]


## Taking Row 2 from dataset

In [31]:
sent, grammar = sentences[index]
sent

'loved it'

In [32]:

# Tokenize the sentence.
tokens = sent.split()

parser = ViterbiParser(grammar)
all_parses = {}

print("---------------------------------------------------------")
print(f"\nsentence: {sent}\nparser: {parser}\ngrammar: {grammar}")
print("---------------------------------------------------------\n\n\n")

---------------------------------------------------------

sentence: loved it
parser: <ViterbiParser for <Grammar with 10 productions>>
grammar: Grammar with 10 productions (start state = S)
    S -> VP NP [0.5]
    S -> VP [0.5]
    VP -> V [0.5]
    VP -> V Det [0.5]
    V -> 'love' [0.5]
    V -> 'loved' [0.5]
    NP -> Det N [1.0]
    Det -> 'my' [0.5]
    Det -> 'it' [0.5]
    N -> 'echo' [1.0]
---------------------------------------------------------





In [33]:
import time

parser.trace(3)
t = time.time()
parses = parser.parse_all(tokens)
time = time.time() - t
average = (
    reduce(lambda a, b: a + b.prob(), parses, 0) / len(parses) if parses else 0
)

Inserting tokens into the most likely constituents table...
   Insert: |=.| loved
   Insert: |.=| it
Finding the most likely constituents spanning 1 text elements...
   Insert: |=.| V -> 'loved' [0.5]                  0.5000000000 
   Insert: |=.| VP -> V [0.5]                       0.2500000000 
   Insert: |=.| S -> VP [0.5]                       0.1250000000 
   Insert: |.=| Det -> 'it' [0.5]                   0.5000000000 
Finding the most likely constituents spanning 2 text elements...
   Insert: |==| VP -> V Det [0.5]                   0.1250000000 
   Insert: |==| S -> VP [0.5]                       0.0625000000 


In [34]:

num_parses = len(parses)
print("Number of parses: ", num_parses)

for p in parses:
    all_parses[p.freeze()] = 1

print("---------------------------------------------------------")
print_stats(all_parses, num_parses)
index = index + 1

Number of parses:  1
---------------------------------------------------------

Time (secs)   # Parses   Average P(parse)
-----------------------------------------
     0.0367          1   0.06250000000000
------------------------------------------
        n/a          1   0.06250000000000


------------------------------------
Printing parses
------------------------------------
(S (VP (V loved) (Det it))) [0.0625]


## Taking Row 3 from dataset

In [35]:
sent, grammar = sentences[index]
sent

'sometimes while playing a game you can answer a question correctly but alexa says you got it wrong and answers the same as you  i like being able to turn lights on and off while away from home'

In [36]:

# Tokenize the sentence.
tokens = sent.split()

parser = ViterbiParser(grammar)
all_parses = {}

print("---------------------------------------------------------")
print(f"\nsentence: {sent}\nparser: {parser}\ngrammar: {grammar}")
print("---------------------------------------------------------\n\n\n")

---------------------------------------------------------

sentence: sometimes while playing a game you can answer a question correctly but alexa says you got it wrong and answers the same as you  i like being able to turn lights on and off while away from home
parser: <ViterbiParser for <Grammar with 78 productions>>
grammar: Grammar with 78 productions (start state = S)
    S -> ADVP NP1 CONJ NP2 CONJ NP3 PNP CONJ NP4 [0.5]
    S -> S1 S2 S3 [0.5]
    S1 -> PN V V DET N ADP N ADP DET N [1.0]
    S2 -> NP5 NP6 CONJ NP7 [1.0]
    S3 -> NP8 ADV NP9 [1.0]
    NP5 -> PN NUM ADV ADJ V ADP N [1.0]
    NP6 -> PN V DET N [1.0]
    NP7 -> V N ADP N [1.0]
    NP8 -> V ADJ N [1.0]
    NP9 -> V N ADV ADV [1.0]
    ADVP -> ADV ADP V DET N [1.0]
    NP1 -> PN V V DET N ADV [1.0]
    NUM -> '4' [1.0]
    NP2 -> N VNP VNP ADJ [1.0]
    VNP -> V PN [1.0]
    NP3 -> V DET ADJ ADP PN [1.0]
    VP -> V V V [1.0]
    NP4 -> ADP ADP ADV ADP N [1.0]
    PNP -> PN VP PRT V N ADP [1.0]
    AD -> ADP ADP [1.0]

In [37]:
import time

parser.trace(3)
t = time.time()
parses = parser.parse_all(tokens)
time = time.time() - t
average = (
    reduce(lambda a, b: a + b.prob(), parses, 0) / len(parses) if parses else 0
)

Inserting tokens into the most likely constituents table...
   Insert: |=.....................................| sometimes
   Insert: |.=....................................| while
   Insert: |..=...................................| playing
   Insert: |...=..................................| a
   Insert: |....=.................................| game
   Insert: |.....=................................| you
   Insert: |......=...............................| can
   Insert: |.......=..............................| answer
   Insert: |........=.............................| a
   Insert: |.........=............................| question
   Insert: |..........=...........................| correctly
   Insert: |...........=..........................| but
   Insert: |............=.........................| alexa
   Insert: |.............=........................| says
   Insert: |..............=.......................| you
   Insert: |...............=......................| got
   Insert: |......

In [38]:

num_parses = len(parses)
print("Number of parses: ", num_parses)

for p in parses:
    all_parses[p.freeze()] = 1

print("---------------------------------------------------------")
print_stats(all_parses, num_parses)
index = index + 1

Number of parses:  1
---------------------------------------------------------

Time (secs)   # Parses   Average P(parse)
-----------------------------------------
     1.8852          1   0.00000000000000
------------------------------------------
        n/a          1   0.00000000000000


------------------------------------
Printing parses
------------------------------------
(S
  (ADVP
    (ADV sometimes)
    (ADP while)
    (V playing)
    (DET a)
    (N game))
  (NP1
    (PN you)
    (V can)
    (V answer)
    (DET a)
    (N question)
    (ADV correctly))
  (CONJ but)
  (NP2
    (N alexa)
    (VNP (V says) (PN you))
    (VNP (V got) (PN it))
    (ADJ wrong))
  (CONJ and)
  (NP3 (V answers) (DET the) (ADJ same) (ADP as) (PN you))
  (PNP
    (PN i)
    (VP (V like) (V being) (V able))
    (PRT to)
    (V turn)
    (N lights)
    (ADP on))
  (CONJ and)
  (NP4
    (ADP off)
    (ADP while)
    (ADV away)
    (ADP from)
    (N home))) [4.786055859375003e-31]


## Taking Row 4 from dataset

In [39]:
sent, grammar = sentences[index]
sent

'i have had a lot of fun with this thing my 4 yr old learns about dinosaurs i control the lights and play games like categories has nice sound when playing music as well'

In [40]:

# Tokenize the sentence.
tokens = sent.split()

parser = ViterbiParser(grammar)
all_parses = {}

print("---------------------------------------------------------")
print(f"\nsentence: {sent}\nparser: {parser}\ngrammar: {grammar}")
print("---------------------------------------------------------\n\n\n")

---------------------------------------------------------

sentence: i have had a lot of fun with this thing my 4 yr old learns about dinosaurs i control the lights and play games like categories has nice sound when playing music as well
parser: <ViterbiParser for <Grammar with 78 productions>>
grammar: Grammar with 78 productions (start state = S)
    S -> ADVP NP1 CONJ NP2 CONJ NP3 PNP CONJ NP4 [0.5]
    S -> S1 S2 S3 [0.5]
    S1 -> PN V V DET N ADP N ADP DET N [1.0]
    S2 -> NP5 NP6 CONJ NP7 [1.0]
    S3 -> NP8 ADV NP9 [1.0]
    NP5 -> PN NUM ADV ADJ V ADP N [1.0]
    NP6 -> PN V DET N [1.0]
    NP7 -> V N ADP N [1.0]
    NP8 -> V ADJ N [1.0]
    NP9 -> V N ADV ADV [1.0]
    ADVP -> ADV ADP V DET N [1.0]
    NP1 -> PN V V DET N ADV [1.0]
    NUM -> '4' [1.0]
    NP2 -> N VNP VNP ADJ [1.0]
    VNP -> V PN [1.0]
    NP3 -> V DET ADJ ADP PN [1.0]
    VP -> V V V [1.0]
    NP4 -> ADP ADP ADV ADP N [1.0]
    PNP -> PN VP PRT V N ADP [1.0]
    AD -> ADP ADP [1.0]
    PRT -> 'to' [1.0]
 

In [41]:
import time

parser.trace(3)
t = time.time()
parses = parser.parse_all(tokens)
time = time.time() - t
average = (
    reduce(lambda a, b: a + b.prob(), parses, 0) / len(parses) if parses else 0
)

Inserting tokens into the most likely constituents table...
   Insert: |=.................................| i
   Insert: |.=................................| have
   Insert: |..=...............................| had
   Insert: |...=..............................| a
   Insert: |....=.............................| lot
   Insert: |.....=............................| of
   Insert: |......=...........................| fun
   Insert: |.......=..........................| with
   Insert: |........=.........................| this
   Insert: |.........=........................| thing
   Insert: |..........=.......................| my
   Insert: |...........=......................| 4
   Insert: |............=.....................| yr
   Insert: |.............=....................| old
   Insert: |..............=...................| learns
   Insert: |...............=..................| about
   Insert: |................=.................| dinosaurs
   Insert: |.................=................| i

In [42]:

num_parses = len(parses)
print("Number of parses: ", num_parses)

for p in parses:
    all_parses[p.freeze()] = 1

print("---------------------------------------------------------")
print_stats(all_parses, num_parses)

Number of parses:  0
---------------------------------------------------------

Time (secs)   # Parses   Average P(parse)
-----------------------------------------
     1.2253          0   0.00000000000000
------------------------------------------
        n/a          0   0.00000000000000
